docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,491 @@
+
+Because HOCs control the definition of the component class itself, they have the power to configure all aspects of a component's behavior, not just what goes in the `render` method.
+
+>**Note:**
+>
+> It's possible to use normal components in place of certain types of HOCs. See the document on render callbacks to learn more about this technique and how it compares.
+
+## Conventions for Idiomatic Higher-Order Components
+
+Although HOCs are a pattern rather than a first-class part of React's API, there are some conventions that we recommend you follow.
+
+### Don't Mutate the Original Component. Use Composition.
+
+Resist the temptation to modify a component's prototype (or otherwise mutate it) inside an HOC.
+
+```js
+function onlyUpdateForKeys(InputComponent, keys) {
+  // Mutates the input component's prototype
+  // *Don't do this!*
+  InputComponent.prototype.shouldComponentUpdate = function(nextProps) {
+    for (let i = 0; i < keys.length; i++) {
+      let key = keys[i];
+      if (
+        !nextProps.hasOwnProperty(key) ||
+        !this.props.hasOwnProperty(key) ||
+        nextProps[key] !== this.props[key]
+      ) {
+        return false;
+      }
+    }
+    return true;
+  }
+  // The fact that we're returning the original input is a hint that it has
+  // been mutated.
+  return InputComponent;
+}
+
+// EnhancedComponent will only update when the `foo` or `bar` props change
+const EnhancedComponent = onlyUpdateForKeys(InputComponent, ['foo', 'bar'])
+```
+
+There are a few problems with this. One is that the input component cannot be reused separately from the enhanced component. More crucially, if you apply another HOC to `EnhancedComponent` that *also* mutates `shouldComponentUpdate`, the first HOC's functionality will be overridden! This HOC also won't work with function components, which do not have lifecycle methods.
+
+Mutating HOCs are a leaky abstraction—the consumer must know how they are implemented in order to avoid conflicts with other HOCs.
+
+Instead of mutation, HOCs should use composition, by wrapping the input component in a container component:
+
+```js
+function onlyUpdateForKeys(WrappedComponent, keys) {
+  return class extends React.Component {
+    shouldComponentUpdate(nextProps) {
+      for (let i = 0; i < keys.length; i++) {
+        let key = keys[i];
+        if (
+          !nextProps.hasOwnProperty(key) ||
+          !this.props.hasOwnProperty(key) ||
+          nextProps[key] !== this.props[key]
+        ) {
+          return false;
+        }
+      }
+      return true;
+    }
+    render() {
+      // Wraps the input component in a container, without mutating it. Good!
+      return <WrappedComponent {...this.props} />;
+    }
+  }
+}
+```
+
+This HOC has the same functionality as the mutating version while avoiding the potential for clashes. It works equally well with class and functional components. And because it's a pure function, it's composable with other HOCs, or even with itself:
```

Comment:
IMO we could just say "Don't mutate the component you're wrapping" in one sentence and leave it at that, rather than providing these code examples.